### PR TITLE
Fix the gathering of importer and populator pods logs

### PR DESCRIPTION
**Pull request by liranr23**
Since we moved to post the VM in the last phase, the VM now usually
contains PVC in the volume spec instead of DV. We need to look on both
in order to gather the relevant data.
Also, adding migrations IDs based on the VM label. This will allow us to
gather the populator pod logs when there is no plan ID provided.

### CODE REVIEW

File: collection-scripts/targeted_logs [hunk 1]
```diff
@@ -84,7 +85,7 @@ for nsvm in ${target_vms[@]}; do
 done
 
 # Collect CDI/importer pod logs for imported DVs
-for nsdv in ${target_dvs[@]}; do
+for nsdv in ${target_dvs[@]} ${target_pvcs[@]}; do
```
Comment (ahadas) [inline collection-scripts/targeted_logs:88]: @liranr23 could you please add few words in the commit message to explain why we do this?

File: collection-scripts/targeted_crs [hunk 2]
```diff
@@ -135,6 +135,13 @@ if [ ! -z "${vm_resources}" ]; then
         dump_resource "virtualmachine" $target_vm_name $target_ns
         present_vm_resources+=("$target_vm_name")
 
+        if [ -z $PLAN ]; then
```
Comment (ahadas) [inline collection-scripts/targeted_crs:138]: what's the assumption here? (if a plan is not specified then .. )

File: collection-scripts/targeted_logs (hunk 1 above)
Comment (liranr23) [reply collection-scripts/targeted_logs:88]: i did have some word about it in the commit.

File: collection-scripts/targeted_crs (hunk 2 above)
Comment (liranr23) [reply collection-scripts/targeted_crs:138]: then, the above migration from the plan didn't set. and you must have either plan or vm+namespace.

File: collection-scripts/targeted_logs (hunk 1 above)
Comment (ahadas) [reply collection-scripts/targeted_logs:88]: ah right but you have two commits so the text doesn't propagate to the description of the PR, I meant the description of the PR...

Comment (liranr23) [reply collection-scripts/targeted_logs:88]: done

File: collection-scripts/targeted_crs [hunk 3]
```diff
@@ -135,6 +135,13 @@ if [ ! -z "${vm_resources}" ]; then
...
+          migration_id=($(echo $vm_data | jq -r '.metadata.labels.migration'))
```
Comment (ahadas) [inline collection-scripts/targeted_crs:139]: ok so you assume here that vm_data was set, which means `if [ ! -z "${vm_resources}" ]; then` so how about checking either vm_data or vm_resources is set rather than checking if the plan wasn't set?

File: collection-scripts/targeted_logs (hunk 1 above)
Comment (ahadas) [inline collection-scripts/targeted_logs:88]: ok, so I understand why you added PVCs here, but we're not going to have importer pods for those PVCs, right? I'm looking at the code below and see things like `$(oc get pods --no-headers -n $ns --selector cdi.kubevirt.io=importer -o go-template='{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}' | grep $dv | head -1)`, will we get something for the PVCs here at all?

Comment (liranr23) [reply collection-scripts/targeted_logs:88]: yes. dv name and pvc name are the same in the case we use dv. the vm contains only pvcs even when using dvs. we do use dvs in case such warm migrations / non local host destination. in that case, the pvc(==dv name) will bring up the right importer pod.

File: collection-scripts/targeted_crs (hunk 2 above)
Comment (ahadas) [reply collection-scripts/targeted_crs:138]: I think we wouldn't fail because the targeted scripts would probably not be called or conditions like `if [ -z $NS ] && [ -z $PLAN ] && [ -z $VM ]` would return false - but need to check
as for the flows you listed above, there seems to be an additional one in which you provide a namespace - need to support that as well or to drop it (but only after we verify it's really not needed)

File: collection-scripts/targeted_logs (hunk 1 above)
Comment (ahadas) [reply collection-scripts/targeted_logs:88]: ok, I didn't say it explicitly but I meant to flows in which we use volume populators and you replied on the other flows.
about the flows you had in mind - ok, you're right that when CDI is used it would be ok to use the PVC names but why having both methods then? can't we reply only on PVC names instead of DV names to get the importer pods?
and about flows in which volume populators are used, I'd prefer not to have logs like `Importer Pod for $nsdv doesn't exist, skipping.` when we don't expect to have importer pods..

File: collection-scripts/targeted_crs (hunk 2 above)
Comment (ahadas) [reply collection-scripts/targeted_crs:138]: > if it's migration then you shold have the vm and namespace

actually you raise a good point here - if we want to work with migration here, why do we still have the option of namespace / plan / vm and don't have one for migration? I would expect the UX to trigger gathering for a plan rather than for a vm and that would render some of the changes in this PR redundant, no?

Comment (liranr23) [reply collection-scripts/targeted_crs:138]: > uld expect the UX to trigger gathering for a migration rather than for a vm and that would render some of the changes in this PR redundant, no?

this works. it gathers everything from the cluster that relates to vms/forklift.

File: collection-scripts/targeted_crs (hunk 3 above)
Comment (liranr23) [reply collection-scripts/targeted_crs:139]: i don't see the benefit in such change. if we have the plan we don't need this block here, as we set already the migration ids based on the plan.

File: collection-scripts/targeted_logs (hunk 1 above)
Comment (liranr23) [reply collection-scripts/targeted_logs:88]: leaving the DVs for me is safety. if the vol-pop will be executed we will get the log you wrote, also for the opposite when passing on the migration and there is no vol-pop pod.

Comment (ahadas) [reply collection-scripts/targeted_logs:88]: > leaving the DVs for me is safety.

can you rephrase / elaborate?

> if the vol-pop will be executed we will get the log you wrote, also for the opposite when passing on the migration and there is no vol-pop pod.

ok, so should we treat both cases differently? that would also solve a syntactic issue with setting `nsdv` to a pvc that has not been originated from a data volume..

Comment (liranr23) [reply collection-scripts/targeted_logs:88]: > > leaving the DVs for me is safety.
> 
> can you rephrase / elaborate?
> 
mainly for supporting older versions.
but thinking on the way we go - we can drop DVs (unless we are going back to them?)
> > if the vol-pop will be executed we will get the log you wrote, also for the opposite when passing on the migration and there is no vol-pop pod.
> 
> ok, so should we treat both cases differently? that would also solve a syntactic issue with setting `nsdv` to a pvc that has not been originated from a data volume..

sure, if we are not going to drop DVs based on the above question

Comment (ahadas) [reply collection-scripts/targeted_logs:88]: we can't drop data volumes for warm migration and we now use them for OCP to OCP - so they're not going to disappear anytime soon

File: collection-scripts/targeted_crs (hunk 3 above)
Comment (liranr23) [reply collection-scripts/targeted_crs]: done with option 2, i need to verify.

Comment (ahadas) [reply collection-scripts/targeted_crs]: yeah, I hoped you'd pick that option :)

File: collection-scripts/targeted_logs [hunk 4]
```diff
@@ -99,6 +100,22 @@ for nsdv in ${target_dvs[@]}; do
   fi
 done
 
+# Collect CDI/importer pod logs for imported PVCs
```
Comment (ahadas) [inline collection-scripts/targeted_logs:103]: can we filter the PVCs that are based on DVs included in target_dvs so we'll operate only on PVCs that we don't expect to have importer pods for?

Comment (liranr23) [reply collection-scripts/targeted_logs:103]: We pass on the VM and look for the volumes. It can be either DV or PVC. We can look into the PVC and understand it came from CDI (and therefore filter it). But it will miss the purpose of getting the importer pods.

The only option to change it is when looking into the vm details if we find a PVC that came from CDI to append it into the DV resources instead of the PVC resources - it feels a bit off to me.

Comment (liranr23) [reply collection-scripts/targeted_logs:103]: I will query the PVC and see it made by CDI. e.g having the label: `app: containerized-data-importer`

Comment (ahadas) [reply collection-scripts/targeted_logs:103]: filtering out those PVCs by looking at their labels sound good, but please try to avoid adding another query - it would be better to use the information that was dumped in dump_resource

Comment (liranr23) [reply collection-scripts/targeted_logs:103]: we don't have the related information we need

Comment (ahadas) [reply collection-scripts/targeted_logs:103]: @liranr23 can you be more specific? you mean that we don't get the PVC with its labels by running `/usr/bin/oc get ${resource} -n ${ocproject} -o yaml ${ocobject}` ?

Comment (liranr23) [reply collection-scripts/targeted_logs:103]: in the dump_resource we dump the pvc name. yes, the command you wrote is another query that will get us what we need

Comment (ahadas) [reply collection-scripts/targeted_logs:103]: ack, yeah so I referred to [that call in dump_resource](https://github.com/kubev2v/forklift-must-gather/blob/main/collection-scripts/targeted_crs#L96)

**Review by ahadas — APPROVED** [review]
